File: plugins/web/src/middleware.rs
//! Declarative middleware for the web app.
//!
//! All cross-cutting HTTP capabilities (CORS, security headers, compression,
//! body limits, JSON API routes) are configured as *data* on the `app` object
//! and applied to the axum router here. This keeps the authoring surface
//! declarative: a GFM table in Marqdo becomes a `middleware` map, and this
//! module turns it into axum layers — 配置即数据、装配即函数.

use std::sync::Arc;

use axum::body::Body;
use axum::extract::State;
use axum::http::{header, HeaderValue, Method, Request, StatusCode};
use axum::middleware::{self as axum_mw, Next};
use axum::response::{IntoResponse, Response};
use axum::routing::{get, post};
use axum::Router;
use serde_json::{json, Map, Value};
use tower_http::compression::CompressionLayer;
use tower_http::cors::{Any, CorsLayer};
use tower_http::limit::RequestBodyLimitLayer;
use tower_http::set_header::SetResponseHeaderLayer;

use crate::db;

/// Render a cell value as text (strings stay, numbers/bools stringify).
fn cell_text(v: &Value) -> String {
    match v {
        Value::Null => String::new(),
        Value::String(s) => s.clone(),
        Value::Bool(b) => b.to_string(),
        Value::Number(n) => n.to_string(),
        other => other.to_string(),
    }
}

/// Parsed middleware configuration (from `app["middleware"]`).
#[derive(Clone, Default)]
pub struct Middleware {
    pub cors: Option<CorsConfig>,
    pub security: Vec<(String, String)>,
    pub compress: bool,
    pub body_limit: Option<u64>,
    /// Log `METHOD path status duration_ms` to stderr for each request.
    pub access_log: bool,
    /// Global `Cache-Control` response header (e.g. `public, max-age=3600`).
    pub cache_control: Option<String>,
    /// JSON API routes: `path -> { method, table, where?, order?, limit? }`.
    pub json_routes: Vec<(String, JsonRoute)>,
}

#[derive(Clone)]
pub struct CorsConfig {
    pub allow_origins: Vec<String>,
    pub methods: Vec<String>,
    pub headers: Vec<String>,
    pub expose_headers: Vec<String>,
    pub credentials: bool,
}

#[derive(Clone)]
pub struct JsonRoute {
    pub method: String,
    pub table: String,
    pub where_v: Option<Value>,
    pub order: Option<String>,
    pub limit: i64,
}

/// Parse `app["middleware"]` (a map) into a `Middleware`.
pub fn parse(app: &Value) -> Middleware {
    let mut mw = Middleware::default();
    let Some(obj) = app.get("middleware").and_then(|v| v.as_object()) else {
        return mw;
    };
    if let Some(c) = obj.get("cors") {
        mw.cors = parse_cors(c);
    }
    if let Some(sec) = obj.get("security").and_then(|v| v.as_object()) {
        mw.security = sec
            .iter()
            .map(|(k, v)| (k.clone(), v.as_str().unwrap_or("").to_string()))
            .filter(|(_, v)| !v.is_empty())
            .collect();
    }
    mw.compress = obj
        .get("compress")
        .and_then(|v| v.as_bool())
        .unwrap_or(false);
    mw.access_log = obj
        .get("access_log")
        .and_then(|v| v.as_bool())
        .unwrap_or(false);
    mw.cache_control = obj
        .get("cache_control")
        .and_then(|v| v.as_str())
        .map(|s| s.trim().to_string())
        .filter(|s| !s.is_empty());
    mw.body_limit = obj
        .get("body_limit")
        .and_then(|v| v.as_u64().or_else(|| v.as_i64().map(|i| i as u64)))
        .filter(|n| *n > 0);
    if let Some(routes) = obj.get("json_routes").and_then(|v| v.as_object()) {
        for (path, spec) in routes {
            if let Some(route) = parse_json_route(path, spec) {
                mw.json_routes.push(route);
            }
        }
        mw.json_routes.sort_by(|a, b| a.0.cmp(&b.0));
    }
    mw
}

fn parse_cors(v: &Value) -> Option<CorsConfig> {
    let obj = v.as_object()?;
    let str_list = |key: &str| -> Vec<String> {
        obj.get(key)
            .and_then(|v| v.as_array())
            .map(|arr| {
                arr.iter()
                    .filter_map(|x| x.as_str().map(|s| s.to_string()))
                    .collect()
            })
            .unwrap_or_default()
    };
    Some(CorsConfig {
        allow_origins: str_list("allow_origins"),
        methods: str_list("methods"),
        headers: str_list("headers"),
        expose_headers: str_list("expose_headers"),
        credentials: obj
            .get("credentials")
            .and_then(|v| v.as_bool())
            .unwrap_or(false),
    })
}

fn parse_json_route(path: &str, spec: &Value) -> Option<(String, JsonRoute)> {
    let spec = spec.as_object()?;
    let method = spec
        .get("method")
        .and_then(|v| v.as_str())
        .unwrap_or("GET")
        .to_uppercase();
    let table = spec.get("table")?.as_str()?.to_string();
    let order = spec.get("order").and_then(|v| v.as_str()).map(|s| s.to_string());
    let limit = spec
        .get("limit")
        .and_then(|v| v.as_i64())
        .unwrap_or(200);
    Some((
        format!("/{}", path.trim_start_matches('/')),
        JsonRoute {
            method,
            table,
            where_v: spec.get("where").cloned(),
            order,
            limit,
        },
    ))
}

/// Apply the middleware configuration to a router (state already provided).
pub fn apply(
    router: Router<Arc<crate::http::AppState>>,
    mw: &Middleware,
) -> Router<Arc<crate::http::AppState>> {
    let mut app = router;

    // JSON API routes first: axum layers only wrap routes registered before
    // the `.layer()` call, so these must be mounted ahead of the pipeline.
    for (path, route) in &mw.json_routes {
        app = mount_json(app, path, route.clone());
    }

    // CORS
    if let Some(c) = &mw.cors {
        let mut layer = CorsLayer::new();
        if c.allow_origins.is_empty() {
            layer = layer.allow_origin(Any);
        } else {
            layer = layer.allow_origin(
                c.allow_origins
                    .iter()
                    .filter_map(|o| o.parse::<HeaderValue>().ok())
                    .collect::<Vec<_>>(),
            );
        }
        if c.methods.is_empty() {
            layer = layer.allow_methods(Any);
        } else {
            layer = layer.allow_methods(
                c.methods
                    .iter()
                    .filter_map(|m| m.parse::<Method>().ok())
                    .collect::<Vec<_>>(),
            );
        }
        if c.headers.is_empty() {
            layer = layer.allow_headers(Any);
        } else {
            layer = layer.allow_headers(
                c.headers
                    .iter()
                    .filter_map(|h| h.parse::<axum::http::HeaderName>().ok())
                    .collect::<Vec<_>>(),
            );
        }
        if !c.expose_headers.is_empty() {
            layer = layer.expose_headers(
                c.expose_headers
                    .iter()
                    .filter_map(|h| h.parse::<axum::http::HeaderName>().ok())
                    .collect::<Vec<_>>(),
            );
        }
        layer = layer.allow_credentials(c.credentials);
        app = app.layer(layer);
    }

    // Security response headers
    for (name, value) in &mw.security {
        let Some(header_name) = header_name_from(&name.to_ascii_lowercase()) else {
            continue;
        };
        if let Ok(v) = HeaderValue::from_str(value) {
            app = app.layer(SetResponseHeaderLayer::overriding(header_name, v));
        }
    }

    // Compression
    if mw.compress {
        app = app.layer(CompressionLayer::new());
    }

    if let Some(ref cc) = mw.cache_control {
        if let Ok(v) = HeaderValue::from_str(cc) {
            app = app.layer(SetResponseHeaderLayer::overriding(
                header::CACHE_CONTROL,
                v,
            ));
        }
    }

    // Request body limit
    if let Some(bytes) = mw.body_limit {
        app = app.layer(RequestBodyLimitLayer::new(bytes as usize));
    }

    // Access log (outermost so status/duration include other layers).
    if mw.access_log {
        app = app.layer(axum_mw::from_fn(access_log_mw));
    }

    app
}

async fn access_log_mw(req: Request<Body>, next: Next) -> Response {
    let method = req.method().clone();
    let path = req.uri().path().to_string();
    let start = std::time::Instant::now();
    let resp = next.run(req).await;
    eprintln!(
        "marqdo web {} {} {} {}ms",
        method,
        path,
        resp.status().as_u16(),
        start.elapsed().as_millis()
    );
    resp
}

fn header_name_from(name: &str) -> Option<axum::http::HeaderName> {
    let h = match name {
        "x_frame_options" | "x-frame-options" => header::X_FRAME_OPTIONS,
        "content_security_policy" | "content-security-policy" => header::CONTENT_SECURITY_POLICY,
        "x_content_type_options" | "x-content-type-options" => header::X_CONTENT_TYPE_OPTIONS,
        "referrer_policy" | "referrer-policy" => header::REFERRER_POLICY,
        "strict_transport_security" | "strict-transport-security" | "hsts" => {
            header::STRICT_TRANSPORT_SECURITY
        }
        "x_xss_protection" | "x-xss-protection" => header::X_XSS_PROTECTION,
        _ => return None,
    };
    Some(h)
}

/// Mount a JSON API route backed by a db query.
fn mount_json(
    app: Router<Arc<crate::http::AppState>>,
    path: &str,
    route: JsonRoute,
) -> Router<Arc<crate::http::AppState>> {
    let handler = match route.method.as_str() {
        "POST" => post(move |State(st): State<Arc<crate::http::AppState>>| {
            json_handler(st, route.clone())
        }),
        _ => get(move |State(st): State<Arc<crate::http::AppState>>| {
            json_handler(st, route.clone())
        }),
    };
    app.route(path, handler)
}

async fn json_handler(st: Arc<crate::http::AppState>, route: JsonRoute) -> Response {
    let Some(url) = st.db_url.as_deref() else {
        return (
            StatusCode::INTERNAL_SERVER_ERROR,
            JsonBody(json!({ "error": "no database" })),
        )
            .into_response();
    };
    let res = match &route.order {
        Some(o) if !o.trim().is_empty() => {
            db::select_order(url, &route.table, route.limit, route.where_v.as_ref(), Some(o), None, None)
        }
        _ => db::select(url, &route.table, route.limit, route.where_v.as_ref()),
    };
    match res {
        Ok(data) => JsonBody(data).into_response(),
        Err(e) => (
            StatusCode::INTERNAL_SERVER_ERROR,
            JsonBody(json!({ "error": e })),
        )
            .into_response(),
    }
}

/// A JSON body response wrapper (sets `Content-Type: application/json`).
pub struct JsonBody(pub Value);

impl IntoResponse for JsonBody {
    fn into_response(self) -> Response {
        Response::builder()
            .status(StatusCode::OK)
            .header(header::CONTENT_TYPE, "application/json; charset=utf-8")
            .body(Body::from(self.0.to_string()))
            .unwrap()
    }
}

/// Normalize a `security` table into the `middleware.security` map shape.
/// Accepts a GFM table `|头|值|` (list of rows) or an already-shaped map.
pub fn security_from_table(table: &Value) -> Value {
    let rows = crate::table::as_rows(table);
    let mut m = Map::new();
    if let Some(arr) = rows.as_array() {
        for row in arr {
            let col = |keys: &[&str]| -> String {
                keys.iter()
                    .find_map(|k| row.get(*k).map(cell_text))
                    .unwrap_or_default()
            };
            let k = col(&["头", "header", "Header", "name"]);
            let v = col(&["值", "value", "Value"]);
            if !k.is_empty() {
                m.insert(k, json!(v));
            }
        }
    }
    Value::Object(m)
}

/// Normalize a CORS table `|允许来源|方法|头|暴露头|凭证|` (one row per origin)
/// into the `middleware.cors` map shape.
pub fn cors_from_table(table: &Value) -> Value {
    let rows = crate::table::as_rows(table);
    let mut allow_origins = Vec::new();
    let mut methods = Vec::new();
    let mut headers = Vec::new();
    let mut expose_headers = Vec::new();
    let mut credentials = false;
    if let Some(arr) = rows.as_array() {
        for row in arr {
            let col = |keys: &[&str]| -> String {
                keys.iter()
                    .find_map(|k| row.get(*k).map(cell_text))
                    .unwrap_or_default()
            };
            let origin = col(&["允许来源", "origin", "Origin"]);
            if !origin.is_empty() && origin != "*" {
                allow_origins.push(origin);
            }
            for m in col(&["方法", "methods", "Methods"]).split(',') {
                let m = m.trim();
                if !m.is_empty() {
                    methods.push(m.to_string());
                }
            }
            for h in col(&["头", "headers", "Headers"]).split(',') {
                let h = h.trim();
                if !h.is_empty() {
                    headers.push(h.to_string());
                }
            }
            for h in col(&["暴露头", "expose_headers", "Expose-Headers"]).split(',') {
                let h = h.trim();
                if !h.is_empty() {
                    expose_headers.push(h.to_string());
                }
            }
            let cred = col(&["凭证", "credentials", "Credentials"]);
            if cred.eq_ignore_ascii_case("true")
                || cred == "1"
                || cred.eq_ignore_ascii_case("yes")
            {
                credentials = true;
            }
        }
    }
    json!({
        "allow_origins": allow_origins,
        "methods": methods,
        "headers": headers,
        "expose_headers": expose_headers,
        "credentials": credentials,
    })
}

/// Normalize a JSON-routes table into `middleware.json_routes` map shape.
/// Accepts a GFM table `|路径|方法|表|条件|排序|上限|` (rows).
pub fn json_routes_from_table(table: &Value) -> Value {
    let rows = crate::table::as_rows(table);
    let mut routes = Map::new();
    if let Some(arr) = rows.as_array() {
        for row in arr {
            let col = |keys: &[&str]| -> String {
                keys.iter()
                    .find_map(|k| row.get(*k).map(cell_text))
                    .unwrap_or_default()
            };
            let path = col(&["路径", "path", "Path"]).trim_start_matches('/').to_string();
            if path.is_empty() {
                continue;
            }
            let method = col(&["方法", "method", "Method"]).to_uppercase();
            let method = if method.is_empty() { "GET".into() } else { method };
            let table = col(&["表", "table", "Table"]);
            if table.is_empty() {
                continue;
            }
            let order = col(&["排序", "order", "Order"]);
            let limit = col(&["上限", "limit", "Limit"]).parse::<i64>().unwrap_or(200);
            let mut spec = Map::new();
            spec.insert("method".into(), json!(method));
            spec.insert("table".into(), json!(table));
            if !order.is_empty() {
                spec.insert("order".into(), json!(order));
            }
            spec.insert("limit".into(), json!(limit));
            routes.insert(path, Value::Object(spec));
        }
    }
    Value::Object(routes)
}

/// Read the `middleware` map out of an app object (for logging/debug).
pub fn summary(app: &Value) -> String {
    let mw = parse(app);
    let mut parts = Vec::new();
    if mw.cors.is_some() {
        parts.push("cors".to_string());
    }
    if !mw.security.is_empty() {
        parts.push("security".to_string());
    }
    if mw.compress {
        parts.push("compress".to_string());
    }
    if mw.body_limit.is_some() {
        parts.push("body_limit".to_string());
    }
    if mw.access_log {
        parts.push("access_log".to_string());
    }
    if mw.cache_control.is_some() {
        parts.push("cache_control".to_string());
    }
    if !mw.json_routes.is_empty() {
        parts.push(format!("json:{}", mw.json_routes.len()));
    }
    if parts.is_empty() {
        String::new()
    } else {
        parts.join(" + ")
    }
}

File: plugins/web/internal/middleware/middleware.go
// Package middleware ports declarative CORS/security/compress/JSON configure
// from the Rust web plugin (配置即数据、装配即函数).
package middleware

import (
	"fmt"
	"strconv"
	"strings"

	"github.com/marqdo/marqdo/plugins/web/internal/table"
)

// Config is the runtime view of app["middleware"].
type Config struct {
	CORS         *CORSConfig
	Security     [][2]string
	Compress     bool
	BodyLimit    *uint64
	AccessLog    bool
	CacheControl string
	JSONRoutes   []JSONRouteMount
}

// CORSConfig matches middleware.cors map shape.
type CORSConfig struct {
	AllowOrigins  []string
	Methods       []string
	Headers       []string
	ExposeHeaders []string
	Credentials   bool
}

// JSONRouteMount is one JSON API route with leading-slash path.
type JSONRouteMount struct {
	Path   string
	Method string
	Table  string
	Where  any
	Order  string
	Limit  int64
}

// Configure merges configure tables into app.middleware (Rust web_app_middleware).
func Configure(appBag map[string]any, args map[string]any) (map[string]any, error) {
	out := clone(appBag)
	mw, _ := out["middleware"].(map[string]any)
	if mw == nil {
		mw = map[string]any{}
	} else {
		mw = clone(mw)
	}

	if v, ok := args["cors"]; ok && v != nil {
		mw["cors"] = CorsFromTable(v)
	}
	if v, ok := args["security"]; ok && v != nil {
		mw["security"] = SecurityFromTable(v)
	}
	if v, ok := args["compress"]; ok && v != nil {
		mw["compress"] = boolish(v)
	}
	if v, ok := first(args, "access_log", "访问日志"); ok {
		mw["access_log"] = boolish(v)
	}
	if v, ok := first(args, "cache_control", "缓存控制"); ok {
		if s := strings.TrimSpace(cellText(v)); s != "" {
			mw["cache_control"] = s
		}
	}
	if v, ok := args["body_limit"]; ok && v != nil {
		if n, ok := asUint64(v); ok && n > 0 {
			mw["body_limit"] = float64(n)
		}
	}
	if v, ok := first(args, "json_routes", "json"); ok && v != nil {
		mw["json_routes"] = JSONRoutesFromTable(v)
	}
	out["middleware"] = mw
	return out, nil
}

// Parse reads app["middleware"] into Config (Rust middleware::parse).
func Parse(app map[string]any) Config {
	var cfg Config
	obj, _ := app["middleware"].(map[string]any)
	if obj == nil {
		return cfg
	}
	if c, ok := obj["cors"].(map[string]any); ok {
		cfg.CORS = parseCORS(c)
	}
	if sec, ok := obj["security"].(map[string]any); ok {
		for k, v := range sec {
			s := cellText(v)
			if s != "" {
				cfg.Security = append(cfg.Security, [2]string{k, s})
			}
		}
	}
	cfg.Compress = boolish(obj["compress"])
	cfg.AccessLog = boolish(obj["access_log"])
	if s := strings.TrimSpace(cellText(obj["cache_control"])); s != "" {
		cfg.CacheControl = s
	}
	if n, ok := asUint64(obj["body_limit"]); ok && n > 0 {
		cfg.BodyLimit = &n
	}
	if routes, ok := obj["json_routes"].(map[string]any); ok {
		for path, spec := range routes {
			if r, ok := parseJSONRoute(path, spec); ok {
				cfg.JSONRoutes = append(cfg.JSONRoutes, r)
			}
		}
		sortJSONRoutes(cfg.JSONRoutes)
	}
	return cfg
}

func parseCORS(obj map[string]any) *CORSConfig {
	return &CORSConfig{
		AllowOrigins:  strList(obj["allow_origins"]),
		Methods:       strList(obj["methods"]),
		Headers:       strList(obj["headers"]),
		ExposeHeaders: strList(obj["expose_headers"]),
		Credentials:   boolish(obj["credentials"]),
	}
}

func parseJSONRoute(path string, spec any) (JSONRouteMount, bool) {
	m, ok := spec.(map[string]any)
	if !ok {
		return JSONRouteMount{}, false
	}
	tableName := strings.TrimSpace(cellText(m["table"]))
	if tableName == "" {
		return JSONRouteMount{}, false
	}
	method := strings.ToUpper(strings.TrimSpace(cellText(m["method"])))
	if method == "" {
		method = "GET"
	}
	limit := int64(200)
	if n, ok := asInt64(m["limit"]); ok {
		limit = n
	}
	order := strings.TrimSpace(cellText(m["order"]))
	p := "/" + strings.TrimPrefix(strings.TrimSpace(path), "/")
	r := JSONRouteMount{
		Path:   p,
		Method: method,
		Table:  tableName,
		Order:  order,
		Limit:  limit,
	}
	if w, ok := m["where"]; ok && w != nil {
		r.Where = w
	}
	return r, true
}

func sortJSONRoutes(rs []JSONRouteMount) {
	for i := 0; i < len(rs); i++ {
		for j := i + 1; j < len(rs); j++ {
			if rs[j].Path < rs[i].Path {
				rs[i], rs[j] = rs[j], rs[i]
			}
		}
	}
}

// CorsFromTable normalizes `|允许来源|方法|头|暴露头|凭证|` into middleware.cors.
func CorsFromTable(tableV any) map[string]any {
	rows, _ := table.AsRows(tableV).([]any)
	var allowOrigins, methods, headers, expose []string
	credentials := false
	for _, row := range rows {
		m, ok := row.(map[string]any)
		if !ok {
			continue
		}
		origin := col(m, "允许来源", "origin", "Origin")
		if origin != "" && origin != "*" {
			allowOrigins = append(allowOrigins, origin)
		}
		for _, part := range strings.Split(col(m, "方法", "methods", "Methods"), ",") {
			part = strings.TrimSpace(part)
			if part != "" {
				methods = append(methods, part)
			}
		}
		for _, part := range strings.Split(col(m, "头", "headers", "Headers"), ",") {
			part = strings.TrimSpace(part)
			if part != "" {
				headers = append(headers, part)
			}
		}
		for _, part := range strings.Split(col(m, "暴露头", "expose_headers", "Expose-Headers"), ",") {
			part = strings.TrimSpace(part)
			if part != "" {
				expose = append(expose, part)
			}
		}
		cred := col(m, "凭证", "credentials", "Credentials")
		if strings.EqualFold(cred, "true") || cred == "1" || strings.EqualFold(cred, "yes") {
			credentials = true
		}
	}
	return map[string]any{
		"allow_origins":  toAnyList(allowOrigins),
		"methods":        toAnyList(methods),
		"headers":        toAnyList(headers),
		"expose_headers": toAnyList(expose),
		"credentials":    credentials,
	}
}

// SecurityFromTable normalizes `|头|值|` into middleware.security map.
func SecurityFromTable(tableV any) map[string]any {
	rows, _ := table.AsRows(tableV).([]any)
	out := map[string]any{}
	for _, row := range rows {
		m, ok := row.(map[string]any)
		if !ok {
			continue
		}
		k := col(m, "头", "header", "Header", "name")
		v := col(m, "值", "value", "Value")
		if k != "" {
			out[k] = v
		}
	}
	return out
}

// JSONRoutesFromTable normalizes `|路径|方法|表|条件|排序|上限|` into json_routes.
func JSONRoutesFromTable(tableV any) map[string]any {
	rows, _ := table.AsRows(tableV).([]any)
	out := map[string]any{}
	for _, row := range rows {
		m, ok := row.(map[string]any)
		if !ok {
			continue
		}
		path := strings.TrimPrefix(strings.TrimSpace(col(m, "路径", "path", "Path")), "/")
		if path == "" {
			continue
		}
		method := strings.ToUpper(strings.TrimSpace(col(m, "方法", "method", "Method")))
		if method == "" {
			method = "GET"
		}
		tbl := col(m, "表", "table", "Table")
		if tbl == "" {
			continue
		}
		order := col(m, "排序", "order", "Order")
		limit := int64(200)
		if s := col(m, "上限", "limit", "Limit"); s != "" {
			if n, err := strconv.ParseInt(s, 10, 64); err == nil {
				limit = n
			}
		}
		spec := map[string]any{
			"method": method,
			"table":  tbl,
			"limit":  float64(limit),
		}
		if order != "" {
			spec["order"] = order
		}
		out[path] = spec
	}
	return out
}

func col(m map[string]any, keys ...string) string {
	for _, k := range keys {
		if v, ok := m[k]; ok {
			return cellText(v)
		}
	}
	return ""
}

func cellText(v any) string {
	switch t := v.(type) {
	case nil:
		return ""
	case string:
		return t
	case bool:
		return strconv.FormatBool(t)
	case float64:
		if t == float64(int64(t)) {
			return strconv.FormatInt(int64(t), 10)
		}
		return strconv.FormatFloat(t, 'f', -1, 64)
	case int:
		return strconv.Itoa(t)
	case int64:
		return strconv.FormatInt(t, 10)
	default:
		return fmt.Sprint(t)
	}
}

func boolish(v any) bool {
	switch t := v.(type) {
	case bool:
		return t
	case string:
		switch strings.TrimSpace(t) {
		case "true", "True", "1", "yes", "on", "真":
			return true
		}
		return false
	case float64:
		return t != 0
	case int:
		return t != 0
	case int64:
		return t != 0
	}
	return false
}

func asUint64(v any) (uint64, bool) {
	switch t := v.(type) {
	case float64:
		if t > 0 {
			return uint64(t), true
		}
	case int64:
		if t > 0 {
			return uint64(t), true
		}
	case int:
		if t > 0 {
			return uint64(t), true
		}
	case string:
		n, err := strconv.ParseUint(strings.TrimSpace(t), 10, 64)
		return n, err == nil && n > 0
	}
	return 0, false
}

func asInt64(v any) (int64, bool) {
	switch t := v.(type) {
	case float64:
		return int64(t), true
	case int64:
		return t, true
	case int:
		return int64(t), true
	case string:
		n, err := strconv.ParseInt(strings.TrimSpace(t), 10, 64)
		return n, err == nil
	}
	return 0, false
}

func strList(v any) []string {
	switch t := v.(type) {
	case []any:
		out := make([]string, 0, len(t))
		for _, x := range t {
			if s, ok := x.(string); ok {
				out = append(out, s)
			}
		}
		return out
	case []string:
		return append([]string{}, t...)
	}
	return nil
}

func toAnyList(ss []string) []any {
	out := make([]any, len(ss))
	for i, s := range ss {
		out[i] = s
	}
	return out
}

func first(args map[string]any, keys ...string) (any, bool) {
	for _, k := range keys {
		if v, ok := args[k]; ok && v != nil {
			return v, true
		}
	}
	return nil, false
}

func clone(m map[string]any) map[string]any {
	out := make(map[string]any, len(m))
	for k, v := range m {
		out[k] = v
	}
	return out
}
